### PR TITLE
[4.0] Require parameter in custom field Subfield

### PR DESCRIPTION
**Pull request by astridx**
Pull Request for Issue # https://github.com/joomla/joomla-cms/issues/27736.

### Summary of Changes
I hid the require parameter in the Custom Field `subfield`.

### Testing Instructions
1. Create a custom field of the type `subfield`
2. See that the parameter require is shown. 
Articles  New Field   test   Administration(1)

3. Apply this patch and see, that the parameter require is not shown anymore.
Articles  New Field   test   Administration

### Expected result
If the option `required` is offered, the value of the field should have an impact. 
In the case of a `subfield`, technically there is always a value. Therefore, the assignment of this option does not matter. 

### Actual result

The field is shown unnecessarily here.

### Documentation Changes Required
Yes
https://docs.joomla.org/Custom_fields_type:_Subfields
https://docs.joomla.org/J3.x:Adding_custom_fields (Here the subfield is missing at all)

### CODE REVIEW

File: plugins/fields/subfields/subfields.php [hunk 1]
```diff
@@ -89,6 +89,10 @@ public function onContentPrepareForm(Form $form, $data)
 		{
 			$valuefield->setAttribute('context', $formData->context);
 		}
+
```
Comment (richard67) [inline plugins/fields/subfields/subfields.php]:
```suggestion

```

File: plugins/fields/subfields/subfields.php [hunk 2]
```diff
@@ -89,6 +89,10 @@ public function onContentPrepareForm(Form $form, $data)
 		{
 			$valuefield->setAttribute('context', $formData->context);
 		}
+		
+		// Don't show the 'required' attribute
```
Comment (richard67) [inline plugins/fields/subfields/subfields.php]:
```suggestion
		// Don't show the 'required' attribute
```

**Review by richard67 — CHANGES_REQUESTED** [review]
Fix PHP code style as reported by drone in PHPCS section.

File: plugins/fields/subfields/subfields.php (hunk 2 above)
Comment (astridx) [reply plugins/fields/subfields/subfields.php]: Done.

File: plugins/fields/subfields/subfields.php (hunk 1 above)
Comment (astridx) [reply plugins/fields/subfields/subfields.php]: Done.

**Comment by richard67** [comment]
@astridx When selecting another field type than "Subflields (subfields)" and setting "required" to "yes" and then changing field type to "Subflields (subfields)" while still editing the field (i.e. not having saved yet), the "required" field is hidden but the value is 1 and not 0. Is that a problem? As far as I understand it is no problem because the field value doesn't have any impact on function for that field type, but I ask just to be sure my understanding is right. @Fedik What do you think?

**Comment by astridx** [comment]
> @richard67 As far as I understand it is no problem because the field value doesn't have any impact on function

Yes, if I have understood everything correctly, the subfield field is saved, regardless of whether require is 1 or 0.
You can only "make" the field as a mandatory field by specifying one `of the field inside the subfield` as a mandatory field.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 4f82d3cbc417fc93829f937fea4745bd0a6678ba<br /><br />Works as promised: When selecting "subfields" as field type, the "required" switcher is hidden, and when selecting another field type the switcher it is shown.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27757">issues.joomla.org/tracker/joomla-cms/27757</a>.</sub>

**Comment by richard67** [comment]
@Fedik Could you test, too?

**Comment by Fedik** [comment]
> When selecting another field type than "Subflields (subfields)" and setting "required" to "yes" and then changing field type to "Subflields (subfields)" while still editing the field (i.e. not having saved yet), the "required" field is hidden but the value is 1 and not 0. Is that a problem?

hm, it sounds like it may have a problem, 
did you tried test it #27736 after that?

I cannot test currently, I try later this week

**Comment by richard67** [comment]
@Fedik Will test later as described in the issue. Thanks for feedback.

**Comment by richard67** [comment]
@Fedik I've just tested. First created a text field with required=no, then created a field where I first had set type type to text field and then had set required=yes and then had changed type to subfields and then saved. then created article and was allowed to save it with emtpy text field. then edited text field and set required to yes, then edited article and verified that i was not allowed to save it because required text field was empty. Finally filled in the required text field and was able to save. So to me it looks all ok.

**Comment by richard67** [comment]
@Fedik Ah, maybe I should have mentioned that what I had found out above was in browser development tools when inspecting the html, not in database. I've tested again as you suggested, and it seems ok to me, see my previous comment.

**Comment by astridx** [comment]
> @Fedik hm, it sounds like it may have a problem

I don't understand where you see the problem. For the `subfield`, it does not matter whether `required` is `1` or `0`. 
Am I missing something?

**Comment by richard67** [comment]
@astridx Maybe he just misunderstood my comment because it was maybe a bit confusing.

**Comment by Fedik** [comment]
@richard67 thanks for checking, btw, it was subform with multiple=yes or multiple=no?

> I don't understand where you see the problem.

It just my assumption, because cannot check on my own currently :wink: 
the idea if you change the field type, then "required=1" still stay, but not editable now,
so I have assume it can lead to this issue #27736

**Comment by richard67** [comment]
@Fedik No, was subform with one text field only. Tested as described [in my comment above](https://github.com/joomla/joomla-cms/pull/27757#issuecomment-581150961).

**Comment by Fedik** [comment]
With multiple=yes or multiple=no I mean this little switcher :wink: 

screen 2020-02-02 19 38 33 424x88

Because it change  behavior a bit.
Need to test both.

**Comment by richard67** [comment]
Ah. Have to test later.

**Comment by Fedik** [comment]
@richard67 I just tested, you were right after changing the field type the value required=1 stays,
however while validation it forced to required=0, and so the issue #27736 somehow fixed 

strange, but okay

**Comment by Fedik** [comment]
I have tested this item :white_check_mark: successfully on 4f82d3cbc417fc93829f937fea4745bd0a6678ba<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27757">issues.joomla.org/tracker/joomla-cms/27757</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27757">issues.joomla.org/tracker/joomla-cms/27757</a>.</sub>

**Comment by richard67** [comment]
> @richard67 I just tested, you were right after changing the field type the value required=1 stays,
> however while validation it forced to required=0, and so the issue #27736 somehow fixed
> 
> strange, but okay

@Fedik Ive seen when I was working on global configuration and installation form for the database connection encryption that this seems to be the way we do. we don't reset fields to default values when hiding them because together with showon conditions and possible default values this can be very complicated. Instead of this we fix that in server side validation in PHP after form validation.

**Comment by rdeutz** [comment]
this dosen't look right to me, removing rtc, I have too look into this again<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27757">issues.joomla.org/tracker/joomla-cms/27757</a>.</sub>

**Comment by astridx** [comment]
As I do not have an idea for a change I close this.

**Pull request closed without merging**